# Contents
    1. Import Libraries
    2. Import Dataframes
    3. Aggregated Mean Calculation for Entire Dataframe
    4. Create a Loyalty Flag
    5. Analyse Spending Habits by Loyalty Category
    6. Create a Spending Flag
    7. Create "median_days_between_orders_column
    8. Create a "order frequency" label flag
    9. Exporting Dataframe

### 1. Import Libraries

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

### 2. Import Dataframe

In [2]:
# create the path

path = r'/Users/vickyczada/Library/CloudStorage/OneDrive-Personal/Documents/Career Foundry/Data Immersion/05-2025 Instacart Basket Analysis'

In [3]:
# importing merged data

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_additional_columns.pkl'))

In [4]:
# check the shape of the dataframe

ords_prods_merge.shape

(32404859, 17)

### 3. Aggregated Mean Calculation for Entire Dataframe

2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [5]:
# grouping by 'department_id' and 'order_number' and calculating the mean- using agg() function

ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

##### How do the results for the entire dataframe differ from the of the subset?

Overall the results of the mean for the entire dataframe are higher across all the departments compared to the subset. This can indicate that while the subset captures enough of the customers behaviours to give us an understanding the entire dataframe will give us a clearer and fuller picture. In both the subset and the entire dataframe department 21 (missing) have the highest mean, highlighting that most of our data is from unknown departments. The rest of the mean from the deparements in the entire dataframe tends to be to the next number i.e. department id 6 (international) was 15.38 in the subset and 16.43 in the entire dataframe. Indicating that while the subset should not be only used it can give similar results, in this instance. 

### 4. Create a Loyalty Flag

4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [6]:
# creating a 'max_order' column

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/bm/wxz2f90s7x104nl7tfs2lwdw0000gn/T/ipykernel_15453/2495908560.py:3: FutureWarning: The provided callable <function max at 0x10673a520> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [7]:
# Checking the Output

ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                      NaN       14084                  2          0   
2                      NaN       12427                  3          0   
3                      NaN       26088                  4          0   
4                      NaN       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

   match    price_range_loc    Busiest days busiest_period_of_day  max_order  
0   both  Mid-range product  Regularly busy        Average orders         10  
1   both  Mid-range product  Regularly busy        Average orders         10  
2   both  Low-range product  Regularly busy        Average orders         10  
3   b

In [8]:
# create a flag that assigns a 'loyalty' label to a user ID

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
# print the frequency of the new 'loyalty_flag'

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [12]:
#  checking the output of first 60 rows only using the “user_id,” “loyalty_flag,” and “order_number” columns

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
5         1      New customer             2
6         1      New customer             2
7         1      New customer             2
8         1      New customer             2
9         1      New customer             2
10        1      New customer             2
11        1      New customer             3
12        1      New customer             3
13        1      New customer             3
14        1      New customer             3
15        1      New customer             3
16        1      New customer             4
17        1      New customer             4
18        1      New customer             4
19        1      New customer             4
20        1      New customer             4
21        1      New customer             5
22        1      New customer             5
23        1      New customer             5
24        1      New customer             5
25        1      New customer             5
26        1      New customer             5
27        1      New customer             5
28        1      New customer             5
29        1      New customer             6
30        1      New customer             6
31        1      New customer             6
32        1      New customer             6
33        1      New customer             7
34        1      New customer             7
35        1      New customer             7
36        1      New customer             7
37        1      New customer             7
38        1      New customer             8
39        1      New customer             8
40        1      New customer             8
41        1      New customer             8
42        1      New customer             8
43        1      New customer             8
44        1      New customer             9
45        1      New customer             9
46        1      New customer             9
47        1      New customer             9
48        1      New customer             9
49        1      New customer             9
50        1      New customer            10
51        1      New customer            10
52        1      New customer            10
53        1      New customer            10
54        1      New customer            10
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1

### 5. Analyse Spending Habits by Loyalty Category

5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [13]:
# Determining Basic Statistics of Spending Habits by Loyalty Category

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Based on the data new customers tend to spend more on average (13.29) compared to loyal (10.39) and regular customers (12.50). The highers average price for new customers could indicate a promotional strategy targeting new users or that they are exploring wider variety of products, including premium items. Whereas loyal customer are spending less on average. Which suggests that they know how to naviagte deals or are more focused on regular purchases rather than exploring rather than the prices of products purchased differing. All of the categories have the same minimum and maximum price, this could be due to consitency across all customer types or a cap on the max prices. This would need to be discussed with the client to determin. 

#### 6. Create a Spending Flag

6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
   
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [14]:
# creating a average product price column

ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/bm/wxz2f90s7x104nl7tfs2lwdw0000gn/T/ipykernel_15453/3615952561.py:3: FutureWarning: The provided callable <function mean at 0x10673af20> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [18]:
# checking relevant columns 

ords_prods_merge[['user_id', 'avg_product_price']].head(100)

user_id  avg_product_price
0         1           6.367797
1         1           6.367797
2         1           6.367797
3         1           6.367797
4         1           6.367797
5         1           6.367797
6         1           6.367797
7         1           6.367797
8         1           6.367797
9         1           6.367797
10        1           6.367797
11        1           6.367797
12        1           6.367797
13        1           6.367797
14        1           6.367797
15        1           6.367797
16        1           6.367797
17        1           6.367797
18        1           6.367797
19        1           6.367797
20        1           6.367797
21        1           6.367797
22        1           6.367797
23        1           6.367797
24        1           6.367797
25        1           6.367797
26        1           6.367797
27        1           6.367797
28        1           6.367797
29        1           6.367797
30        1           6.367797
31        1           6.367797
32        1           6.367797
33        1           6.367797
34        1           6.367797
35        1           6.367797
36        1           6.367797
37        1           6.367797
38        1           6.367797
39        1           6.367797
40        1           6.367797
41        1           6.367797
42        1           6.367797
43        1           6.367797
44        1           6.367797
45        1           6.367797
46        1           6.367797
47        1           6.367797
48        1           6.367797
49        1           6.367797
50        1           6.367797
51        1           6.367797
52        1           6.367797
53        1           6.367797
54        1           6.367797
55        1           6.367797
56        1           6.367797
57        1           6.367797
58        1           6.367797
59        2           7.515897
60        2           7.515897
61        2           7.515897
62        2           7.515897
63        2           7.515897
64        2           7.515897
65        2           7.515897
66        2           7.515897
67        2           7.515897
68        2           7.515897
69        2           7.515897
70        2           7.515897
71        2           7.515897
72        2           7.515897
73        2           7.515897
74        2           7.515897
75        2           7.515897
76        2           7.515897
77        2           7.515897
78        2           7.515897
79        2           7.515897
80        2           7.515897
81        2           7.515897
82        2           7.515897
83        2           7.515897
84        2           7.515897
85        2           7.515897
86        2           7.515897
87        2           7.515897
88        2           7.515897
89        2           7.515897
90        2           7.515897
91        2           7.515897
92        2           7.515897
93        2           7.515897
94        2           7.515897
95        2           7.515897
96        2           7.515897
97        2           7.515897
98        2           7.515897
99        2           7.515897

In [17]:
pd.options.display.max_rows = None

In [21]:
# Creating a 'Spending' Label Flag using loc Function

ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [22]:
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [23]:
# Examining Value Count on 'spending_flag' Column

ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [24]:
# Checking Releavent Columns

ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0         1           6.367797   Low spender
1         1           6.367797   Low spender
2         1           6.367797   Low spender
3         1           6.367797   Low spender
4         1           6.367797   Low spender
5         1           6.367797   Low spender
6         1           6.367797   Low spender
7         1           6.367797   Low spender
8         1           6.367797   Low spender
9         1           6.367797   Low spender
10        1           6.367797   Low spender
11        1           6.367797   Low spender
12        1           6.367797   Low spender
13        1           6.367797   Low spender
14        1           6.367797   Low spender

7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
   
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

### 7. Create "median_days_between_orders_column

In [26]:
# create 'median_days_between_orders' column

ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/bm/wxz2f90s7x104nl7tfs2lwdw0000gn/T/ipykernel_15453/162984200.py:3: FutureWarning: The provided callable <function median at 0x106c5dbc0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [27]:
# Checking Relevant Columns 

ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0         1                     NaN                        20.5
1         1                     NaN                        20.5
2         1                     NaN                        20.5
3         1                     NaN                        20.5
4         1                     NaN                        20.5
5         1                    15.0                        20.5
6         1                    15.0                        20.5
7         1                    15.0                        20.5
8         1                    15.0                        20.5
9         1                    15.0                        20.5
10        1                    15.0                        20.5
11        1                    21.0                        20.5
12        1                    21.0                        20.5
13        1                    21.0                        20.5
14        1                    21.0                        20.5

### 8. Create a "order frequency" label flag

In [28]:
# Creating a 'Order Frequency' Label Flag using loc Function

ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [31]:
# Examining Value Counts on 'order_frequency_flag'

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [32]:
# Checking Relevant Columns of Updated Dataframe

ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     NaN                        20.5   
1         1                     NaN                        20.5   
2         1                     NaN                        20.5   
3         1                     NaN                        20.5   
4         1                     NaN                        20.5   
5         1                    15.0                        20.5   
6         1                    15.0                        20.5   
7         1                    15.0                        20.5   
8         1                    15.0                        20.5   
9         1                    15.0                        20.5   
10        1                    15.0                        20.5   
11        1                    21.0                        20.5   
12        1                    21.0                        20.5   
13        1                    21.0                        20.5   
14        1                    21.0                        20.5   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10  Non-frequent customer  
11  Non-frequent customer  
12  Non-frequent customer  
13  Non-frequent customer  
14  Non-frequent customer

In [33]:
# Exporting the Dataframe

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))